### PR TITLE
adds support for FQCN form type referencing to FormContractor

### PR DESCRIPTION
**Pull request by ju1ius**
I am targetting this branch, because this is a BC bugfix

Will close https://github.com/sonata-project/SonataAdminBundle/issues/3976 when [doctrine-orm-admin-bundle PR](https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/598) is also merged.
## Changelog

``` markdown
### Fixed
- Fix `FormContractor::getDefaultOptions` not checking against form types FQCNs
```
## Subject

`FormContractor::getDefaultOptions` didn't check against form types FQCNs, causing the default options not to be set when using the Symfony 3 way:

``` php
$formMapper->add('foo', ModelType::class);
```

See related admin-bundle issue at https://github.com/sonata-project/SonataAdminBundle/issues/3976

### CODE REVIEW

File: Builder/FormContractor.php [hunk 1]
```diff
@@ -84,31 +84,52 @@ public function getDefaultOptions($type, FieldDescriptionInterface $fieldDescrip
...
 
             $options['class'] = $fieldDescription->getTargetEntity();
             $options['model_manager'] = $fieldDescription->getAdmin()->getModelManager();
-        } elseif ($type == 'sonata_type_admin') {
+
+            if ($type === 'sonata_type_model_autocomplete'
+                || $type === 'Sonata\AdminBundle\Form\Type\ModelAutocompleteType'
+            ) {
+                if (!$fieldDescription->getAssociationAdmin()) {
+                    throw new \RuntimeException(sprintf('The current field `%s` is not linked to an admin. Please create one for the target entity: `%s`', $fieldDescription->getName(), $fieldDescription->getTargetEntity()));
```
Comment (OskarStark) [inline Builder/FormContractor.php]: please linebreak this

File: Builder/FormContractor.php [hunk 2]
```diff
@@ -84,31 +84,52 @@ public function getDefaultOptions($type, FieldDescriptionInterface $fieldDescrip
...
+            'Sonata\AdminBundle\Form\Type\ModelType',
+            'sonata_type_model_list',
+            'Sonata\AdminBundle\Form\Type\ModelTypeList',
+            'sonata_type_model_hidden',
+            'Sonata\AdminBundle\Form\Type\ModelHiddenType',
+            'sonata_type_model_autocomplete',
+            'Sonata\AdminBundle\Form\Type\ModelAutocompleteType',
+        ), true)) {
+            if ($fieldDescription->getOption('edit') === 'list') {
                 throw new \LogicException('The ``sonata_type_model`` type does not accept an ``edit`` option anymore, please review the UPGRADE-2.1.md file from the SonataAdminBundle');
```
Comment (OskarStark) [inline Builder/FormContractor.php]: please linebreak this

File: Builder/FormContractor.php [hunk 3]
```diff
@@ -84,31 +84,52 @@ public function getDefaultOptions($type, FieldDescriptionInterface $fieldDescrip
...
+            if ($type === 'sonata_type_model_autocomplete'
+                || $type === 'Sonata\AdminBundle\Form\Type\ModelAutocompleteType'
+            ) {
+                if (!$fieldDescription->getAssociationAdmin()) {
+                    throw new \RuntimeException(sprintf('The current field `%s` is not linked to an admin. Please create one for the target entity: `%s`', $fieldDescription->getName(), $fieldDescription->getTargetEntity()));
+                }
+            }
+        } elseif ($type === 'sonata_type_admin' || $type === 'Sonata\AdminBundle\Form\Type\AdminType') {
             if (!$fieldDescription->getAssociationAdmin()) {
                 throw new \RuntimeException(sprintf('The current field `%s` is not linked to an admin. Please create one for the target entity : `%s`', $fieldDescription->getName(), $fieldDescription->getTargetEntity()));
```
Comment (OskarStark) [inline Builder/FormContractor.php]: please linebreak this

File: Builder/FormContractor.php [hunk 4]
```diff
@@ -84,31 +84,52 @@ public function getDefaultOptions($type, FieldDescriptionInterface $fieldDescrip
...
+            // set sensitive default value to have a component working fine out of the box
+            $options['btn_add'] = false;
+            $options['delete'] = false;
+
             $options['data_class'] = $fieldDescription->getAssociationAdmin()->getClass();
             $fieldDescription->setOption('edit', $fieldDescription->getOption('edit', 'admin'));
-        } elseif ($type == 'sonata_type_collection') {
+        } elseif ($type === 'sonata_type_collection' || $type === 'Sonata\CoreBundle\Form\Type\CollectionType') {
             if (!$fieldDescription->getAssociationAdmin()) {
                 throw new \RuntimeException(sprintf('The current field `%s` is not linked to an admin. Please create one for the target entity : `%s`', $fieldDescription->getName(), $fieldDescription->getTargetEntity()));
```
Comment (OskarStark) [inline Builder/FormContractor.php]: please linebreak this

File: Builder/FormContractor.php (hunk 1 above)
Comment (ju1ius) [inline Builder/FormContractor.php]: @OskarStark I didn't change this line, that's why I didn't line-break it.

File: Builder/FormContractor.php (hunk 2 above)
Comment (ju1ius) [inline Builder/FormContractor.php]: Same as above

File: Builder/FormContractor.php (hunk 3 above)
Comment (ju1ius) [inline Builder/FormContractor.php]: Same as above

File: Builder/FormContractor.php (hunk 4 above)
Comment (ju1ius) [inline Builder/FormContractor.php]: Same as above

File: Builder/FormContractor.php (hunk 1 above)
Comment (OskarStark) [inline Builder/FormContractor.php]: i understand

@greg0ire what do you think?

Comment (greg0ire) [inline Builder/FormContractor.php]: I think it should be left to the contributor's discretion.
@ju1ius if you feel like doing it, please do so in a separate commit.

File: Builder/FormContractor.php [hunk 5]
```diff
@@ -84,31 +84,53 @@ public function getDefaultOptions($type, FieldDescriptionInterface $fieldDescrip
...
+            'sonata_type_model',
+            'Sonata\AdminBundle\Form\Type\ModelType',
+            'sonata_type_model_list',
+            'Sonata\AdminBundle\Form\Type\ModelTypeList',
+            'sonata_type_model_hidden',
+            'Sonata\AdminBundle\Form\Type\ModelHiddenType',
+            'sonata_type_model_autocomplete',
+            'Sonata\AdminBundle\Form\Type\ModelAutocompleteType',
+        ), true)) {
+            if ($fieldDescription->getOption('edit') === 'list') {
```
Comment (Soullivaneuh) [inline Builder/FormContractor.php]: Yoda condition is needed according to Symfony conventions.

File: Builder/FormContractor.php [hunk 6]
```diff
@@ -84,31 +84,53 @@ public function getDefaultOptions($type, FieldDescriptionInterface $fieldDescrip
...
                 throw new \LogicException('The ``sonata_type_model`` type does not accept an ``edit`` option anymore, please review the UPGRADE-2.1.md file from the SonataAdminBundle');
             }
 
             $options['class'] = $fieldDescription->getTargetEntity();
             $options['model_manager'] = $fieldDescription->getAdmin()->getModelManager();
-        } elseif ($type == 'sonata_type_admin') {
+
+            if ($type === 'sonata_type_model_autocomplete'
+                || $type === 'Sonata\AdminBundle\Form\Type\ModelAutocompleteType'
```
Comment (Soullivaneuh) [inline Builder/FormContractor.php]: Yoda condition is needed according to Symfony conventions.

Comment (Soullivaneuh) [inline Builder/FormContractor.php]: Or why not using `in_array`?

File: Builder/FormContractor.php [hunk 7]
```diff
@@ -84,31 +84,53 @@ public function getDefaultOptions($type, FieldDescriptionInterface $fieldDescrip
...
+            ) {
+                if (!$fieldDescription->getAssociationAdmin()) {
+                    throw new \RuntimeException(sprintf('The current field `%s` is not linked to an admin. Please create one for the target entity: `%s`', $fieldDescription->getName(), $fieldDescription->getTargetEntity()));
+                }
+            }
+        } elseif ($type === 'sonata_type_admin' || $type === 'Sonata\AdminBundle\Form\Type\AdminType') {
```
Comment (Soullivaneuh) [inline Builder/FormContractor.php]: `in_array`

File: Builder/FormContractor.php [hunk 8]
```diff
@@ -84,31 +84,53 @@ public function getDefaultOptions($type, FieldDescriptionInterface $fieldDescrip
...
             }
 
+            // set sensitive default value to have a component working fine out of the box
+            $options['btn_add'] = false;
+            $options['delete'] = false;
+
             $options['data_class'] = $fieldDescription->getAssociationAdmin()->getClass();
             $fieldDescription->setOption('edit', $fieldDescription->getOption('edit', 'admin'));
-        } elseif ($type == 'sonata_type_collection') {
+        } elseif ($type === 'sonata_type_collection' || $type === 'Sonata\CoreBundle\Form\Type\CollectionType') {
```
Comment (Soullivaneuh) [inline Builder/FormContractor.php]: `in_array`

File: Tests/Builder/FormContractorTest.php [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    protected function setUp()
+    {
+        $this->formFactory = $this->getMock('Symfony\Component\Form\FormFactoryInterface');
+
+        $this->formContractor = new FormContractor($this->formFactory);
+    }
+
+    public function testDefaultOptionsForSonataFormTypes()
+    {
+        $admin = $this->getMockBuilder('Sonata\AdminBundle\Admin\AdminInterface')->getMock();
```
Comment (Soullivaneuh) [inline Tests/Builder/FormContractorTest.php]: Use `getMock` for interfaces.

File: Tests/Builder/FormContractorTest.php [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        $modelManager = $this->getMockBuilder('Sonata\AdminBundle\Model\ModelManagerInterface')->getMock();
```
Comment (Soullivaneuh) [inline Tests/Builder/FormContractorTest.php]: Use `getMock` for interfaces.

File: Tests/Builder/FormContractorTest.php [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        $modelClass = 'FooEntity';
+
+        $admin->method('getModelManager')->willReturn($modelManager);
+        $admin->method('getClass')->willReturn($modelClass);
+
+        $fieldDescription = $this->getMockBuilder('Sonata\AdminBundle\Admin\FieldDescriptionInterface')->getMock();
```
Comment (Soullivaneuh) [inline Tests/Builder/FormContractorTest.php]: Use `getMock` for interfaces.

File: Builder/FormContractor.php (hunk 5 above)
Comment (Soullivaneuh) [inline Builder/FormContractor.php]: No need to just vote :-1: because you don't like it. It's a convention so please follow it.

File: Builder/FormContractor.php [hunk 12]
```diff
@@ -84,31 +84,53 @@ public function getDefaultOptions($type, FieldDescriptionInterface $fieldDescrip
...
             }
 
             $options['class'] = $fieldDescription->getTargetEntity();
             $options['model_manager'] = $fieldDescription->getAdmin()->getModelManager();
-        } elseif ($type == 'sonata_type_admin') {
+
+            if ($type === 'sonata_type_model_autocomplete'
+                || $type === 'Sonata\AdminBundle\Form\Type\ModelAutocompleteType'
+            ) {
+                if (!$fieldDescription->getAssociationAdmin()) {
```
Comment (Soullivaneuh) [inline Builder/FormContractor.php]: If forest detected. :-)

File: Builder/FormContractor.php [hunk 13]
```diff
@@ -84,31 +84,53 @@ public function getDefaultOptions($type, FieldDescriptionInterface $fieldDescrip
...
+        ), true)) {
+            if ($fieldDescription->getOption('edit') === 'list') {
                 throw new \LogicException('The ``sonata_type_model`` type does not accept an ``edit`` option anymore, please review the UPGRADE-2.1.md file from the SonataAdminBundle');
             }
 
             $options['class'] = $fieldDescription->getTargetEntity();
             $options['model_manager'] = $fieldDescription->getAdmin()->getModelManager();
-        } elseif ($type == 'sonata_type_admin') {
+
+            if ($type === 'sonata_type_model_autocomplete'
```
Comment (Soullivaneuh) [inline Builder/FormContractor.php]: Why this new `if`? Is that related to this PR?

**Comment by core23** [comment]
Please rebase to retrigger StyleCI @ju1ius

File: Builder/FormContractor.php (hunk 6 above)
Comment (joshlopes) [inline Builder/FormContractor.php]: `Sonata\AdminBundle\Form\Type\ModelAutocompleteType` instead use ModelAutocompleteType::class ? @Soullivaneuh

Comment (greg0ire) [inline Builder/FormContractor.php]: Yes b/c we are BC-compatible with old php versions

**Comment by core23** [comment]
Are you still working on this @ju1ius ?

**Comment by OskarStark** [comment]
any news here @ju1ius ?

**Comment by greg0ire** [comment]
@OskarStark I'm afraid you won't get any news from @ju1ius … he has not been having any activity on github for a long time. That's a bit worrying to be honest :worried:

**Comment by OskarStark** [comment]
> That's a bit worrying to be honest 😟

😢

**Comment by jordisala1991** [comment]
This was already done on #252
